File: CODE_SNIPPETS/TESTING/test_examples.py
"""
Test Examples for REChain Code Snippets
Pytest-style test examples for testing bridge components
"""

import asyncio
import pytest
from unittest.mock import Mock, AsyncMock, patch
from typing import Dict, List


class TestMatrixClient:
    """Tests for Matrix client functionality."""
    
    def test_client_config_creation(self):
        """Test creating client configuration."""
        config = {
            'homeserver_url': 'https://matrix.example.com',
            'user_id': '@test:example.com',
            'access_token': 'test_token'
        }
        
        assert config['homeserver_url'] is not None
        assert config['user_id'].startswith('@')
        assert config['access_token'] is not None
    
    def test_room_id_format(self):
        """Test room ID format validation."""
        room_id = '!room123:example.com'
        
        assert room_id.startswith('!')
        assert ':' in room_id
        assert len(room_id) > 10
    
    def test_event_content_structure(self):
        """Test event content structure."""
        content = {
            'msgtype': 'm.text',
            'body': 'Hello World',
            'formatted_body': '<b>Hello World</b>',
            'format': 'org.matrix.custom.html'
        }
        
        assert content['msgtype'] == 'm.text'
        assert 'body' in content
        assert isinstance(content['body'], str)
    
    def test_user_id_validation(self):
        """Test user ID validation."""
        def is_valid_user_id(user_id: str) -> bool:
            return user_id.startswith('@') and ':' in user_id
        
        assert is_valid_user_id('@user:example.com')
        assert not is_valid_user_id('user:example.com')
        assert not is_valid_user_id('@user')
    
    def test_message_types(self):
        """Test message type constants."""
        MESSAGE_TYPES = {
            'text': 'm.text',
            'image': 'm.image',
            'video': 'm.video',
            'audio': 'm.audio',
            'file': 'm.file',
            'location': 'm.location',
            'emote': 'm.emote',
            'notice': 'm.notice'
        }
        
        assert len(MESSAGE_TYPES) == 8
        assert MESSAGE_TYPES['text'] == 'm.text'
        assert MESSAGE_TYPES['emote'] == 'm.emote'


class TestBridgeConfig:
    """Tests for bridge configuration."""
    
    def test_rate_limit_config(self):
        """Test rate limit configuration."""
        rate_limit = {
            'enabled': True,
            'requests_per_second': 50.0,
            'burst_size': 100
        }
        
        assert rate_limit['enabled'] is True
        assert rate_limit['requests_per_second'] > 0
        assert rate_limit['burst_size'] >= rate_limit['requests_per_second']
    
    def test_bridge_room_mapping(self):
        """Test bridge room mapping."""
        room_mapping = {
            'telegram': {
                'chat_id': '-100123456',
                'matrix_room_id': '!room123:example.com'
            },
            'discord': {
                'channel_id': '123456789',
                'matrix_room_id': '!room456:example.com'
            }
        }
        
        assert 'telegram' in room_mapping
        assert 'discord' in room_mapping
        assert 'matrix_room_id' in room_mapping['telegram']
    
    def test_database_config(self):
        """Test database configuration."""
        db_config = {
            'host': 'localhost',
            'port': 5432,
            'database': 'bridge',
            'user': 'bridge_user',
            'pool_size': 10,
            'max_overflow': 20
        }
        
        assert db_config['port'] > 0
        assert db_config['pool_size'] > 0
        assert db_config['max_overflow'] >= db_config['pool_size']
    
    def test_security_config(self):
        """Test security configuration."""
        security = {
            'encryption': {
                'enabled': True,
                'algorithm': 'AES-256-GCM'
            },
            'rate_limiting': {
                'enabled': True,
                'max_requests': 100
            }
        }
        
        assert security['encryption']['enabled'] is True
        assert security['rate_limiting']['enabled'] is True


class TestCircuitBreaker:
    """Tests for circuit breaker functionality."""
    
    def test_circuit_states(self):
        """Test circuit breaker states."""
        states = ['CLOSED', 'OPEN', 'HALF_OPEN']
        
        assert len(states) == 3
        assert 'CLOSED' in states
        assert 'OPEN' in states
        assert 'HALF_OPEN' in states
    
    def test_failure_threshold(self):
        """Test failure threshold handling."""
        config = {
            'failure_threshold': 5,
            'timeout': 60.0
        }
        
        failures = 0
        for i in range(config['failure_threshold']):
            failures += 1
        
        assert failures == config['failure_threshold']
    
    def test_circuit_transitions(self):
        """Test circuit state transitions."""
        transitions = {
            ('CLOSED', 'OPEN'): 'Failure threshold reached',
            ('OPEN', 'HALF_OPEN'): 'Timeout expired',
            ('HALF_OPEN', 'CLOSED'): 'Success threshold reached',
            ('HALF_OPEN', 'OPEN'): 'Failure in half-open'
        }
        
        assert len(transitions) == 4
        assert ('CLOSED', 'OPEN') in transitions


class TestRateLimiter:
    """Tests for rate limiter functionality."""
    
    def test_token_bucket(self):
        """Test token bucket algorithm."""
        config = {
            'rate': 10.0,      # tokens per second
            'capacity': 20     # max tokens
        }
        
        tokens = config['capacity']
        requests = 5
        
        for _ in range(requests):
            if tokens > 0:
                tokens -= 1
        
        assert tokens == config['capacity'] - requests
    
    def test_rate_limit_exceeded(self):
        """Test rate limit exceeded condition."""
        rate_limit = {
            'max_requests': 100,
            'window_seconds': 60
        }
        
        requests = [True] * 100 + [False] * 10
        
        allowed = sum(1 for r in requests if r) <= rate_limit['max_requests']
        
        assert allowed is True
    
    def test_burst_handling(self):
        """Test burst handling."""
        burst_limit = 50
        burst = 50
        
        assert burst <= burst_limit
        assert burst > 0


class TestMetrics:
    """Tests for metrics functionality."""
    
    def test_counter_increment(self):
        """Test counter increment."""
        counter = 0
        
        for _ in range(10):
            counter += 1
        
        assert counter == 10
    
    def test_histogram_buckets(self):
        """Test histogram bucket configuration."""
        buckets = [0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0]
        
        assert len(buckets) == 10
        assert buckets[0] < buckets[-1]
        assert all(buckets[i] < buckets[i+1] for i in range(len(buckets)-1))
    
    def test_gauge_value(self):
        """Test gauge value setting."""
        gauge = {
            'rooms_bridged': 5,
            'users_connected': 100,
            'queue_size': 10
        }
        
        assert gauge['rooms_bridged'] >= 0
        assert gauge['users_connected'] >= 0
        assert gauge['queue_size'] >= 0


class TestAsyncPatterns:
    """Tests for async patterns."""
    
    def test_async_queue(self):
        """Test async queue operations."""
        queue = []
        
        for i in range(5):
            queue.append(i)
        
        item = queue.pop(0)
        
        assert item == 0
        assert len(queue) == 4
    
    def test_task_creation(self):
        """Test async task creation."""
        async def dummy_task():
            return 'done'
        
        tasks = [dummy_task() for _ in range(3)]
        
        assert len(tasks) == 3
    
    def test_cancellation(self):
        """Test task cancellation."""
        cancelled = False
        
        # Simulate cancellation
        is_cancelled = True
        
        if is_cancelled:
            cancelled = True
        
        assert cancelled is True


class TestValidation:
    """Tests for validation utilities."""
    
    def test_message_length_validation(self):
        """Test message length validation."""
        max_length = 65536
        message = 'x' * 100
        
        assert len(message) < max_length
    
    def test_url_validation(self):
        """Test URL format validation."""
        def is_valid_mxc(url: str) -> bool:
            return url.startswith('mxc://') and len(url) > 10
        
        assert is_valid_mxc('mxc://example.com/file123')
        assert not is_valid_mxc('https://example.com/file')
    
    def test_json_parsing(self):
        """Test JSON parsing."""
        import json
        
        data = '{"type": "m.room.message", "content": {}}'
        parsed = json.loads(data)
        
        assert parsed['type'] == 'm.room.message'
        assert isinstance(parsed['content'], dict)


# Run tests if executed directly
if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: CODE_SNIPPETS/TESTING/test_examples.js
/**
 * Test Examples for REChain Code Snippets
 * 
 * Provides pytest-style test examples for testing bridge components.
 */

const assert = require('assert');

// Test utilities
class TestSuite {
    constructor(name) {
        this.name = name;
        this.tests = [];
        this.beforeEachHooks = [];
        this.afterEachHooks = [];
        this.beforeAllHooks = [];
        this.afterAllHooks = [];
    }

    beforeEach(fn) {
        this.beforeEachHooks.push(fn);
    }

    afterEach(fn) {
        this.afterEachHooks.push(fn);
    }

    test(name, fn) {
        this.tests.push({ name, fn });
    }

    async run() {
        console.log(`\n${'='.repeat(60)}`);
        console.log(`Test Suite: ${this.name}`);
        console.log(`${'='.repeat(60)}`);

        let passed = 0;
        let failed = 0;
        let skipped = 0;

        for (const test of this.tests) {
            const ctx = {
                skip: (msg) => { throw new Error('SKIP:' + msg); }
            };

            try {
                // Run beforeEach hooks
                for (const hook of this.beforeEachHooks) {
                    await hook(ctx);
                }

                await test.fn(ctx);

                console.log(`✅ ${test.name}`);
                passed++;
            } catch (e) {
                if (e.message.startsWith('SKIP:')) {
                    console.log(`⏭️  ${test.name} (SKIP: ${e.message.slice(5)})`);
                    skipped++;
                } else {
                    console.log(`❌ ${test.name}`);
                    console.log(`   Error: ${e.message}`);
                    failed++;
                }
            }
        }

        console.log(`\nResults: ${passed} passed, ${failed} failed, ${skipped} skipped`);
        return failed === 0;
    }
}

// Example: Testing Matrix Client
const matrixTests = new TestSuite('Matrix Client Tests');

matrixTests.test('should create client with valid config', () => {
    const config = {
        homeserverUrl: 'https://matrix.example.com',
        userId: '@test:example.com',
        accessToken: 'test_token'
    };
    assert(config.homeserverUrl !== undefined);
    assert(config.userId !== undefined);
});

matrixTests.test('should parse room ID correctly', () => {
    const roomId = '!room123:example.com';
    assert(roomId.startsWith('!'));
    assert(roomId.includes(':'));
});

matrixTests.test('should validate message format', () => {
    const message = {
        msgtype: 'm.text',
        body: 'Hello World',
        format: 'org.matrix.custom.html'
    };
    assert(message.msgtype === 'm.text');
    assert(typeof message.body === 'string');
});

matrixTests.test('should handle event types', () => {
    const eventTypes = [
        'm.room.message',
        'm.room.encrypted',
        'm.room.member',
        'm.room.name'
    ];
    assert(eventTypes.length === 4);
    assert(eventTypes.includes('m.room.message'));
});

// Example: Testing Bridge Configuration
const bridgeTests = new TestSuite('Bridge Configuration Tests');

bridgeTests.test('should validate bridge config', () => {
    const config = {
        homeserverUrl: 'https://matrix.example.com',
        appserviceToken: '__APPSERVICE_TOKEN__',
        homeserverToken: '__HS_TOKEN__',
        bridgeName: 'telegram',
        botUserId: '@telegram_bot:example.com'
    };
    
    assert(config.appserviceToken !== '__CHANGE_ME__');
    assert(config.homeserverToken !== '__CHANGE_ME__');
    assert(config.bridgeName.length > 0);
});

bridgeTests.test('should validate rate limit config', () => {
    const rateLimit = {
        enabled: true,
        requestsPerSecond: 50,
        burstSize: 100
    };
    
    assert(rateLimit.enabled === true);
    assert(rateLimit.requestsPerSecond > 0);
    assert(rateLimit.burstSize >= rateLimit.requestsPerSecond);
});

bridgeTests.test('should validate room mapping', () => {
    const roomMapping = {
        'telegram': {
            'chatId': '!matrixRoom:example.com'
        }
    };
    
    assert(typeof roomMapping.telegram === 'object');
    assert(roomMapping.telegram.chatId !== undefined);
});

// Example: Testing Circuit Breaker
const circuitBreakerTests = new TestSuite('Circuit Breaker Tests');

circuitBreakerTests.test('should start in closed state', () => {
    const state = 'CLOSED';
    assert(state === 'CLOSED');
});

circuitBreakerTests.test('should transition to open after failures', () => {
    const config = { failureThreshold: 5 };
    let failures = 0;
    
    for (let i = 0; i < config.failureThreshold; i++) {
        failures++;
    }
    
    assert(failures === config.failureThreshold);
});

circuitBreakerTests.test('should allow recovery after timeout', async () => {
    const state = 'OPEN';
    const timeout = 60;
    
    assert(timeout > 0);
    
    // Simulate timeout
    await new Promise(resolve => setTimeout(resolve, 100));
    const newState = 'HALF_OPEN';
    
    assert(newState === 'HALF_OPEN');
});

// Run tests
async function runAllTests() {
    console.log('\n' + '='.repeat(60));
    console.log('REChain Code Snippets - Test Suite');
    console.log('='.repeat(60) + '\n');
    
    let allPassed = true;
    
    allPassed &= await matrixTests.run();
    allPassed &= await bridgeTests.run();
    allPassed &= await circuitBreakerTests.run();
    
    console.log('\n' + '='.repeat(60));
    if (allPassed) {
        console.log('✅ All tests passed!');
    } else {
        console.log('❌ Some tests failed');
        process.exit(1);
    }
    console.log('='.repeat(60));
}

module.exports = { TestSuite, runAllTests };

// Run if executed directly
if (require.main === module) {
    runAllTests();
}

